trieste/models/gpflow/models.py

```diff
@@ -1647,3 +1647,367 @@ def covariance_with_top_fidelity(self, query_points: TensorType) -> TensorType:
+
+            sample_mean = tf.where(mask, next_fidelity_sample_mean, sample_mean)  # [..., N, 1, S]
+            sample_var = tf.where(mask, next_fidelity_sample_var, sample_var)  # [..., N, 1, S]
+
+        return sample_mean, sample_var
+
+    def _propogate_samples_through_level(
+        self,
+        query_point: TensorType,
+        fidelity: int,
+        sample_mean: TensorType,
+        sample_var: TensorType,
+    ) -> tuple[TensorType, TensorType]:
+        """
+        Propogate samples through a given fidelity.
+
+        This takes a set of query points without a fidelity column and calculates samples
+        at the given fidelity, using the sample means and variances from the previous fidelity.
+
+        :param query_points: The query points to sample at, with no fidelity column,
+            with shape[..., N, D]
+        :param fidelity: The fidelity to propogate the samples through
+        :param sample_mean: Samples of the posterior mean at the previous fidelity,
+            with shape [..., N, 1, S]
+        :param sample_var: Samples of the posterior variance at the previous fidelity,
+            with shape [..., N, 1, S]
+        :return: sample_mean: Samples of the posterior mean at the given fidelity,
+            of shape [..., N, 1, S]
+            and sample_var: Samples of the posterior variance at the given fidelity,
+            of shape [..., N, 1, S]
+        """
+        B, N, D = query_point.shape[:-2], query_point.shape[-2], query_point.shape[-1]
+        S = sample_mean.shape[-1]
+        # Repeat random numbers for each query point and add middle dimension
+        # for concatentation with query points This also means that it has the same
+        # shape as sample_var and sample_mean, so there's no broadcasting required
+        # Note: at the moment we use the same monte carlo values for every value in the batch dim
+        reshaped_random_numbers = tf.repeat(
+            tf.transpose(self.monte_carlo_random_numbers), N, axis=-2
+        )[
+            ..., None, :
+        ]  # [..., N, 1, S]
+        samples = reshaped_random_numbers * tf.sqrt(sample_var) + sample_mean  # [..., N, 1, S]
+        # Add an extra unit dim to query_point and repeat for each of the samples
+        qp_repeated = tf.repeat(
+            query_point[..., :, :, None],  # [..., N, D, 1]
+            self.monte_carlo_random_numbers.shape[0],
+            axis=-1,
+        )  # [..., N, D, S]
+        qp_augmented = tf.concat([qp_repeated, samples], axis=-2)  # [..., N, D+1, S]
+
+        # Flatten sample dimension to n_qp dimension to pass through predictor
+        # Switch dims to make reshape match up correct dimensions for query points
+        # Use Einsum to switch last two dimensions
+        qp_augmented = tf.linalg.matrix_transpose(qp_augmented)  # [..., N, S, D+1]
+        flat_qp_augmented = tf.reshape(qp_augmented, shape=[*B, N * S, D + 1])  # [..., N*S, D+1]
```

Comment:
we have `flatten_leading_dims` here https://github.com/secondmind-labs/trieste/blob/develop/trieste/utils/misc.py#L240

not sure if it would be useful?

Reply:
Yes, that seems to do what I need it to, thanks!